### PR TITLE
Refactor CowrieExtractionStrategy and add comprehensive tests

### LINKED ISSUE #634: Refactor CowrieExtractionStrategy
The CowrieExtractionStrategy class (currently only present in `develop`) contains a lot of legacy code which is not very testable. It should be considered to refactor this class and write a comprehensive set of tests for it.

### PR DESCRIPTION
**Pull request by opbot-xd**
# Description

Refactors the [CowrieExtractionStrategy](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie.py:13:0-255:19) to decouple log parsing from database operations, making the logic granularly testable. This addresses issue #634 and lays the groundwork for end-to-end pipeline tests (#636).

1.  **New [CowrieLogParser](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie_parser.py:25:0-144:23) Class**: Created a standalone parser in [greedybear/cronjobs/extraction/strategies/cowrie_parser.py](cci:7://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie_parser.py:0:0-0:0). This class handles all logic for extracting payloads, downloads, and sessions from raw hits without any database dependencies.
2.  **Refactored Strategy**: Updated [CowrieExtractionStrategy](cci:2://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie.py:13:0-255:19) to act as a coordinator. It now delegates parsing to the new helper and focuses solely on persistence and linking IOCs.
3.  **Regex Fix**: Fixed `REGEX_URL_PROTOCOL` in [greedybear/regex.py](cci:7://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/regex.py:0:0-0:0) to correctly support URLs with protocols (e.g., `http://...`), which were previously missed if the colon was present.
4.  **Tests**: Added two new test suites ([tests/greedybear/cronjobs/test_cowrie_parser.py](cci:7://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/tests/greedybear/cronjobs/test_cowrie_parser.py:0:0-0:0) and [tests/greedybear/cronjobs/test_cowrie_strategy.py](cci:7://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/tests/greedybear/cronjobs/test_cowrie_strategy.py:0:0-0:0)) verifying parser logic and strategy flows.

## Related issues
Closes #634 and #644

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [ ] I have added documentation of the new features.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### CODE REVIEW

**Comment by opbot-xd** [comment]
**Fix:** The original regex `ps?` only matched protocols followed immediately by `//` (e.g. `http//`), missing standard URLs like `http://`. 
I added `:?` to optionally match the colon, ensuring we capture valid URLs found in payloads (as confirmed by the new unit tests) while maintaining backward compatibility for loose matches.

**Comment by opbot-xd** [comment]
@regulartim please review the PR when you get time.
Also I noticed that flake8-django occasionally crashes with an AttributeError related to astroid when parsing newer Python syntax (like dataclasses or type aliases). I was able to work around this in my PR by simplifying the code, so my checks are passing now.

However, intended to mention that this seems to be due to flake8-django requiring an older version of astroid (<3.0) while the CI environment currently pulls the latest version (4.x). Pinning astroid<3.0.0 in `requirements-linters.txt` might help prevent this friction for future contributions.

**Comment by mlodic** [comment]
actually, there's a problem with flake8-django which is not maintained anymore. We could definitively consider switching all our linters to Ruff https://github.com/astral-sh/ruff. I'll open an issue

**Comment by regulartim** [comment]
> Also I noticed that flake8-django occasionally crashes with an AttributeError related to astroid when parsing newer Python syntax (like dataclasses or type aliases). I was able to work around this in my PR by simplifying the code, so my checks are passing now.
However, intended to mention that this seems to be due to flake8-django requiring an older version of astroid (<3.0) while the CI environment currently pulls the latest version (4.x). Pinning astroid<3.0.0 in requirements-linters.txt might help prevent this friction for future contributions.

See #640

**Comment by regulartim** [comment]
> @regulartim please review the PR when you get time.

Yes, I will review an test later today.

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        self.source_ip = source_ip
+        self.start_time: Any = None
+        self.duration: float = 0.0
+        self.credentials: list[str] = []
+        self.commands: list[str] = []
+        self.interaction_count: int = 0
+        self.login_attempt: bool = False
+        self.command_execution: bool = False
+        self.commands_first_seen: Any = None
+        self.commands_last_seen: Any = None
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: This looks quite similar to the CowrieSession model from `models.py`. What is the reason you are not using that?

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                        results.append(
+                            {
+                                "source_ip": str(hit.get("src_ip")),
+                                "download_url": url,
+                                "hostname": hostname,
+                            }
+                        )
+                except ValueError:
+                    self.log.warning(f"Failed to parse URL {url}")
+                    continue
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: This is extremely similar to the url parsing in `extract_payloads`. Consider putting this logic into a separate function.

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                elif eventid == "cowrie.command.input":
+                    session_data.command_execution = True
+                    if session_data.commands_first_seen is None:
+                        session_data.commands_first_seen = timestamp
+
+                    message = hit.get("message", "")
+                    if message.startswith("CMD: "):
+                        command = message[5:].replace("\x00", "[NUL]")
+                    else:
+                        command = message.replace("\x00", "[NUL]")
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: Can you explain why you need to handle two cases here?

**Comment by regulartim** [comment]
I won't finish the review today. Sorry.

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
+import re
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: I think the file does not belong into `greedybear/cronjobs/extraction/strategies/` but in `greedybear/cronjobs/extraction/` as it is not a strategy but a extraction related helper. Do you agree?

**Comment by regulartim** [comment]
> Fix: The original regex ps? only matched protocols followed immediately by // (e.g. http//), missing standard URLs like http://.
I added :? to optionally match the colon, ensuring we capture valid URLs found in payloads (as confirmed by the new unit tests) while maintaining backward compatibility for loose matches.

Good catch! :+1: Can you open an issue for that? Just for documenting it.

**Comment by opbot-xd** [comment]
> Good catch! 👍 Can you open an issue for that? Just for documenting it.


opened #644 for this

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py (hunk 3 above)
Comment (regulartim) [reply greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: The original code uses `removeprefix()`. Is anything wrong with that?

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+        # Group hits by session first
+        hits_per_session = {}
+        for hit in hits:
+            sid = hit.get("session")
+            if not sid:
+                continue
+            if sid not in hits_per_session:
+                hits_per_session[sid] = []
+            hits_per_session[sid].append(hit)
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: Why don't you use a defaultdict as in the original code? Makes it much cleaner, in my opinion.

File: greedybear/cronjobs/extraction/strategies/cowrie_parser.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+            # Find source IP
+            src_ip = None
+            for h in sorted_hits:
+                if "src_ip" in h:
+                    src_ip = h["src_ip"]
+                    break
+
+            if not src_ip:
+                continue
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie_parser.py]: You can savely assume that there is a "src_ip" in the hit. Hits without this filed are filtered in `greedybear/cronjobs/extraction/pipeline.py`

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 7]
```diff
@@ -52,48 +65,51 @@ def _get_scanners(self, hits: list[dict]) -> None:
...
-                continue
-            if hit.get("eventid", "") != "cowrie.session.file_download":
-                continue
-            self.log.info(f"found IP {hit['src_ip']} trying to execute download from {hit['url']}")
-            scanner_ip = str(hit["src_ip"])
-            ioc = IOC(name=scanner_ip, type=get_ioc_type(scanner_ip), cowrie=True)
-            ioc_record = self.ioc_processor.add_ioc(ioc, attack_type=SCANNER)
-            if ioc_record:
+                # Note: Previous code called payload/session extraction here
+                # We now do it separately in extract_from_hits
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Remove that comment please. It has no value to future readers.

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 8]
```diff
@@ -106,50 +122,95 @@ def _get_url_downloads(self, hits: list[dict]) -> None:
...
                 self.session_repo.save_command_sequence(session_record.commands)
-                self.log.info(f"saved new command execute from {ioc.name} " f"with hash {session_record.commands.commands_hash}")
+                self.log.info(f"saved new command execute from {source_ip} " f"with hash {session_record.commands.commands_hash}")
+
             self.ioc_repo.save(session_record.source)
             self.session_repo.save_session(session_record)
 
-        self.log.info(f"{len(hits_per_session)} sessions added")
-
     def _add_fks(self, scanner_ip, hostname):
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Please also refactor this method. It's one of the main points where we can simplify the code

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 9]
```diff
@@ -106,50 +122,95 @@ def _get_url_downloads(self, hits: list[dict]) -> None:
...
-                        session_record.duration = hit["duration"]
-                session_record.interaction_count += 1
+    def _save_sessions(self, sessions: Dict[str, CowrieSessionData]) -> None:
+        for sid, session_data in sessions.items():
+            source_ip = session_data.source_ip
+            # We need the IOC object for the source IP
+            # It should have been created by _get_scanners
+            scanner_ioc = self.ioc_repo.get_ioc_by_name(source_ip)
+            if not scanner_ioc:
+                # Should not happen if data consistency holds, but fallback:
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: I think I would throw an exception or at least log an error and continue. Because at this point _get_scanners() already finished, and if there is no IOC something in the code went horribly wrong, right?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 10]
```diff
@@ -106,50 +122,95 @@ def _get_url_downloads(self, hits: list[dict]) -> None:
...
+            # If I append session_data.credentials to session_record.credentials, fine.
+            # Wait, `session_record.credentials` is ArrayField (list).
+            # I should probably just set it if I assume I have the full view of the session?
+            # No, maybe only partial hits.
+            # Safety: append.
+
+            # Wait, `session_record.credentials` default is list.
+            if session_data.credentials:
+                # Original used append one by one.
+                session_record.credentials.extend(session_data.credentials)
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: This thinking process does not belong in the code as a comment. Also this code is plain wrong: if session_data holds credentials, you are extending twice.

Comment (regulartim) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: That said, I am open to discuss deduplicating credentials.

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 11]
```diff
@@ -106,50 +122,95 @@ def _get_url_downloads(self, hits: list[dict]) -> None:
...
+
+            if session_data.login_attempt:
+                # Need to increment login_attempts on source IOC?
+                # Original: session_record.source.login_attempts += 1 (inside loop for each hit)
+                # So calculate total attempts from session_data
+                count = len(session_data.credentials)
+                # Or based on hits? Original: for hit in hits: if connect/failed: login_attempts += 1.
+                # session_data.credentials corresponds to failed/success hits.
+                session_record.source.login_attempts += count
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Same, that's a thinking process that does not belong into the code. Also why not just `session_record.source.login_attempts = len(session_data.credentials)`? You don't need the condition!

**Comment by opbot-xd** [comment]
Hi @regulartim,

Thank you for the detailed feedback! I've completely refactored the code based on your suggestions.

## Changes Made

**Structural:**
- Consolidated `cowrie_parser.py` into `cowrie.py` (single file)
- Removed `CowrieSessionData` class - work directly with models
- Extracted helper functions: `parse_url_hostname()`, `normalize_command()`, `normalize_credential_field()`

**Code Quality:**
- Removed unnecessary comments
- Added comprehensive docstrings (100% coverage)
- Used `defaultdict(list)` and `removeprefix()` as suggested
- Simplified `_add_fks` with early return

**Testing:**
- Added 25 unit tests covering all functions and edge cases
- All 265 tests passing (was 240)

## About Those Comments...

I forgot to remove them before committing. Basically left my scratch notes in production code.

Same with that bad `ioc_arg.type` vs `attack_type` assertion in tests - completely different fields (IOC type vs attack classification). Forgot to clean up that debugging check too.

## Apology

Sorry for the messy initial PR. Learnt a lot from your feedback and will be more careful in the future.

Let me know if anything else needs improvement!

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 12]
```diff
@@ -115,81 +210,105 @@ def _get_sessions(self, ioc: IOC, hits: list[dict]) -> None:
...
+
         cmd_seq = self.session_repo.get_command_sequence_by_hash(commands_hash=commands_hash)
         if cmd_seq is None:
-            # In case sequence does not exist:
-            # Assign hash to the the sequence
             session.commands.commands_hash = commands_hash
             return False
-        # In case sequence does already exist:
-        # Delete newly created sequence from DB
-        # and assign existing sequence to session
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py:190]: Why did you remove these comments?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 13]
```diff
@@ -115,81 +210,105 @@ def _get_sessions(self, ioc: IOC, hits: list[dict]) -> None:
...
+        """
         scanner_ip_instance = self.ioc_repo.get_ioc_by_name(scanner_ip)
         hostname_instance = self.ioc_repo.get_ioc_by_name(hostname)
 
-        if scanner_ip_instance is not None:
-            if hostname_instance and hostname_instance not in scanner_ip_instance.related_ioc.all():
-                scanner_ip_instance.related_ioc.add(hostname_instance)
+        # Early return if either IOC doesn't exist
+        if not scanner_ip_instance or not hostname_instance:
+            return
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py:273]: Nice! If one of them does not exist, does that indicate a problem with the extraction process? What do you think? If so, it would be better to write an error to the log, right?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 14]
```diff
@@ -115,81 +210,105 @@ def _get_sessions(self, ioc: IOC, hits: list[dict]) -> None:
...
+
+        # Link scanner -> hostname
+        if hostname_instance not in scanner_ip_instance.related_ioc.all():
+            scanner_ip_instance.related_ioc.add(hostname_instance)
             self.ioc_repo.save(scanner_ip_instance)
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Django doc says that "Using add() on a relation that already exists won’t duplicate the relation, but it will still trigger signals."
https://docs.djangoproject.com/en/6.0/ref/models/relations/#django.db.models.fields.related.RelatedManager.add

That's why I think we don't need the if-condition at all, right?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 15]
```diff
@@ -115,81 +210,105 @@ def _get_sessions(self, ioc: IOC, hits: list[dict]) -> None:
...
 
-        if hostname_instance is not None:
-            if scanner_ip_instance and scanner_ip_instance not in hostname_instance.related_ioc.all():
-                hostname_instance.related_ioc.add(scanner_ip_instance)
+        # Link hostname -> scanner
+        if scanner_ip_instance not in hostname_instance.related_ioc.all():
+            hostname_instance.related_ioc.add(scanner_ip_instance)
             self.ioc_repo.save(hostname_instance)
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Same here!

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 16]
```diff
@@ -56,44 +110,78 @@ def _get_scanners(self, hits: list[dict]) -> None:
...
-            ioc_record = self.ioc_processor.add_ioc(ioc, attack_type=SCANNER)
+            download_url = str(hit["url"])
+
+            self.log.info(f"found IP {scanner_ip} trying to execute download from {download_url}")
+
+            # Track scanner IP
+            scanner_ioc = IOC(name=scanner_ip, type=get_ioc_type(scanner_ip), cowrie=True)
+            ioc_record = self.ioc_processor.add_ioc(scanner_ioc, attack_type=SCANNER)
             if ioc_record:
                 self.added_ip_downloads += 1
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: Didn't we already add this IoC in the `_get_scanners` method?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 17]
```diff
@@ -56,44 +110,78 @@ def _get_scanners(self, hits: list[dict]) -> None:
                 self._get_sessions(ioc_record, hits)
 
     def _extract_possible_payload_in_messages(self, scanner_ip: str, hits: list[dict]) -> None:
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py]: This method is called for each scanner IoC and loops through all the hits. Do you think there is a more efficient solution?

File: greedybear/cronjobs/extraction/strategies/cowrie.py [hunk 18]
```diff
@@ -13,12 +14,57 @@
...
+    """
+    Normalize command string by removing CMD prefix and null characters.
+
+    Args:
+        message: Raw command message string
+
+    Returns:
+        Normalized command string, truncated to 1024 characters
+    """
+    return message.removeprefix("CMD: ").replace("\x00", "[NUL]")[:1024]
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/strategies/cowrie.py:45]: Maybe also document why we do the 1024 character truncation. It's because the how the filed is defined in the model: `commands = pg_fields.ArrayField(models.CharField(max_length=1024, blank=True), blank=False, null=False, default=list)`

File: tests/test_cowrie_extraction.py [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        result = normalize_credential_field("user\x00name")
+        self.assertEqual(result, "user[NUL]name")
+
+    def test_normalize_credential_field_clean(self):
+        """Test credential normalization with clean string."""
+        result = normalize_credential_field("admin")
+        self.assertEqual(result, "admin")
+
+
+class TestCowrieExtractionStrategy(TestCase):
```
Comment (regulartim) [inline tests/test_cowrie_extraction.py:73]: Please add test where `hits` is empty. This is not covered, as far as I can see.

**Review by regulartim — CHANGES_REQUESTED** [review]
I think this is a step in the right direction. If I am asking questions in my review, it's not because I necessarily want you to change things, but because I want to discuss the code with you. I'm not an experienced software engineer and often times I'm just wrong. So I invite you to answer my questions and find good solutions together.

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 12 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py:190]: I felt that the code itself explained it by proper variable names so should remove it.

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 13 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py:273]: Good catch!
**Current behavior:**
The early return happens when we can't find one of the IOCs we just created moments earlier.
**When would this happen?**
Under normal operation, it shouldn't - we only call [_add_fks()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie.py:267:4-290:49) immediately after `add_ioc()` succeeds. If the IOCs don't exist, it likely indicates:
- IOC processor didn't save properly
- Database transaction issue
- Race condition (unlikely)

I'll add a warning log. Here's what I'll change:

```python
def _add_fks(self, scanner_ip: str, hostname: str) -> None:
    scanner_ip_instance = self.ioc_repo.get_ioc_by_name(scanner_ip)
    hostname_instance = self.ioc_repo.get_ioc_by_name(hostname)

    # Log if IOCs are missing - this shouldn't happen in normal operation
    if not scanner_ip_instance or not hostname_instance:
        self.log.warning(
            f"Cannot link IOCs - missing from database: "
            f"scanner_ip={scanner_ip_instance is not None}, "
            f"hostname={hostname_instance is not None}"
        )
        return

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 14 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: Yes you are right infact My `not in related_ioc.all()` forces a database query every time.
Will remove it.

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 12 above)
Comment (regulartim) [reply greedybear/cronjobs/extraction/strategies/cowrie.py:190]: Ok, I mean you are the one that saw the code for the first time a few days ago. If you think the code explains itself, we can omit the comment.

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 16 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: I was trying to ensure we track which IPs performed downloads specifically, but `ioc_processor.add_ioc()` is idempotent - it won't create duplicates. So the scanner was already added in step 1.
But you're absolutely right! The scanner IP is already added in [_get_scanners()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie.py:99:4-109:52) which runs first. The `ioc_processor.add_ioc()` call here is redundant.
I'll remove the scanner tracking (lines 171-176) and just keep:
1. Parse download URL hostname
2. Create payload IOC
3. Link them via [_add_fks()](cci:1://file:///home/opbotxd/Desktop/vscode/dev/gsoc/honeynet/GreedyBear/greedybear/cronjobs/extraction/strategies/cowrie.py:267:4-292:45)
The counter `added_ip_downloads` was also misleading since these IPs were already added earlier.

Comment (regulartim) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: Cool! :+1:  If we also have logging in _add_fks(), we will know if our assumption here is false.

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 17 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: **Better approach:**
Process all hits once, group by src_ip:
```python
def _extract_possible_payload_in_messages(self, hits: list[dict]) -> None:
    # Group payloads by scanner IP in one pass
    payloads_by_ip = defaultdict(list)
    for hit in hits:  # Single M iteration
        if hit.get("eventid") not in ["cowrie.login.failed", "cowrie.session.file_upload"]:
            continue
        match_url = REGEX_URL_PROTOCOL.search(hit.get("message", ""))
        if match_url:
            payloads_by_ip[hit["src_ip"]].append(match_url.group())
    
    # Then process grouped payloads
    for scanner_ip, urls in payloads_by_ip.items():
        for url in urls:
            # Create payload IOC and link
```


This reduces complexity from O(N*M) to O(M) + O(P) where P = total payloads found, N = scanners, M = hits.

Is this correct?

File: tests/test_cowrie_extraction.py (hunk 19 above)
Comment (opbot-xd) [reply tests/test_cowrie_extraction.py:73]: Added test for empty hits:
```python
def test_extract_payload_in_messages_empty_hits(self):
    """Test extraction with empty hits list."""
    hits = []
    self.strategy._extract_possible_payload_in_messages(hits)
    self.strategy.ioc_processor.add_ioc.assert_not_called()

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 18 above)
Comment (opbot-xd) [reply greedybear/cronjobs/extraction/strategies/cowrie.py:45]: Adding this comment
```python
    # Truncate to 1024 chars to match CommandSequence.commands field max_length
    return message.removeprefix("CMD: ").replace("\x00", "[NUL]")[:1024]

File: greedybear/cronjobs/extraction/strategies/cowrie.py (hunk 17 above)
Comment (regulartim) [reply greedybear/cronjobs/extraction/strategies/cowrie.py]: I think it is. :+1:
